BUILD.bazel

```diff
@@ -0,0 +1,379 @@
+    outs = ["frontends/parsers/p4/p4lexer.lex"],
+    cmd = ("sed '/%option outfile=\"lex.yy.c\"/d' $< > $@"),
+    visibility = ["//visibility:private"],
+)
+
+genlex(
+    name = "p4lexer",
+    src = "frontends/parsers/p4/p4lexer.lex",
+    out = "frontends/parsers/p4/p4lexer.cc",
+    prefix = "yy",
+    visibility = ["//visibility:private"],
+)
+
+genrule(
+    name = "v1lexer_lex",
+    srcs = ["frontends/parsers/v1/v1lexer.ll"],
+    outs = ["frontends/parsers/v1/v1lexer.lex"],
+    cmd = ("sed '/%option outfile=\"lex.yy.c\"/d' $< > $@"),
+    visibility = ["//visibility:private"],
+)
+
+genlex(
+    name = "v1lexer",
+    src = "frontends/parsers/v1/v1lexer.lex",
+    out = "frontends/parsers/v1/v1lexer.cc",
+    prefix = "yy",
+    visibility = ["//visibility:private"],
+)
+
+cc_library(
+    name = "p4c_includes",
+    includes = [
+        "ir/",
+        "lib/",
+        "tools/ir-generator",
+    ],
+    visibility = [":__subpackages__"],
+)
+
+# This rule helps split some circular dependencies between subdirectories.
+cc_library(
+    name = "p4c_frontend_h",
+    hdrs = [
+        "frontends/p4/typeChecking/typeSubstitution.h",
+    ] + glob([
+        "frontends/p4/*.h",
+    ]) + glob([
+        "frontends/common/*.h",
+    ]),
+    visibility = ["//visibility:private"],
+)
+
+# The ir-generator tool uses a parser built by these genlex and genyacc rules.
+genlex(
+    name = "ir_generator_lex",
+    src = "tools/ir-generator/ir-generator-lex.l",
+    out = "tools/ir-generator/ir-generator-lex.c",
+    includes = [
+        "tools/ir-generator/ir-generator-yacc.hh",
+    ],
+    prefix = "yy",
+)
+
+genyacc(
+    name = "ir_generator_yacc",
+    src = "tools/ir-generator/ir-generator.ypp",
+    header_out = "tools/ir-generator/ir-generator-yacc.hh",
+    source_out = "tools/ir-generator/ir-generator-yacc.cc",
+)
+
+# This cc_library contains the ir-generator tool sources, including the
+# ir-generator parser source from the lex/yacc output. The srcs attribute
+# excludes generator.cpp since it is part of the cc_binary.
+cc_library(
+    name = "p4c_ir_generator_lib",
+    srcs = [
+        "tools/ir-generator/ir-generator-yacc.cc",
+    ] + glob(
+        ["tools/ir-generator/*.cpp"],
+        exclude = ["tools/ir-generator/generator.cpp"],
+    ),
+    hdrs = [
+        "tools/ir-generator/ir-generator-lex.c",
+        "tools/ir-generator/ir-generator-yacc.hh",
+    ] + glob([
+        "tools/ir-generator/*.h",
+    ]),
+    deps = [
+        ":p4c_includes",
+        ":p4c_toolkit",
+    ],
+)
+
+# The next rule builds the ir-generator tool binary.
+cc_binary(
+    name = "irgenerator",
+    srcs = ["tools/ir-generator/generator.cpp"],
+    linkopts = ["-lgmp"],
+    visibility = [":__subpackages__"],
+    deps = [
+        ":p4c_ir_generator_lib",
+        ":p4c_toolkit",
+    ],
+)
+
+genrule(
+    name = "ir_generated_files",
+    srcs = [
+        "frontends/p4-14/ir-v1.def",
+        "backends/bmv2/bmv2.def",
```

Comment:
Could this be factored so that the backend and the frontend are not mingled together?
Ideally the backend should be treated as an add-on that depends on the frontend.
I..e, if we add a new backend we should not need to edit the main makefile, esp. if the backend is an extension.
This is not a must, but would be nice to have.

Reply:
Are you suggestion that we invoke the ir generator twice? Is that possible? If so I'm happy to refactor things that way if you'd prefer.